### PR TITLE
feat: Timepoint Explorer

### PR DESCRIPTION
**Pull request by ckbedwell**
TimepointExplorer

## Timepoint Explorer

https://github.com/grafana/synthetic-monitoring-app/blob/feat/custom-logs/docs/timepointExplorer/index.md

**☝️ Strongly advised pre-read ☝️**

It is finally here! The Timepoint Explorer is a bespoke data visualisation component in the Synthetic Monitoring application to easily view the individual executions that your checks are reporting as well as understanding many different events which accompany your executions.

## Notes for reviewing

Highly suggested to read the documentation linked above. It not only helps to explain all the different features that the Timepoint Explorer supports but it currently acts as a reference when updating the Timepoint Explorer in the future what the expected behaviour is.

Normally for a feature like this I'd be writing tons of integration tests however in the interest of time and getting this live, I am favouring more unit tests and relying on documentation to act as a stop gap to this blindspot. The main issue with writing integration tests for the Timepoint Explorer is it has several interconnected API calls that need mocking via MSW, so essentially need a bespoke testing framework so we have maximum confidence. This may be better suited for when we have an e2e testing suite (which we also lack).

**This PR isn't suitable to merge to `main` directly yet. It also has dependencies on:**

- [x] Convert all Scenes dashboards to React Scenes
- [x] Move key metrics above Timepoint Explorer in all dashboards
- [x] Instrument the Timepoint Explorer with analytics events
- [x] Implement feedback component to help gather qualitative feedback of its impact

With that said, this PR is ready for review as intended. Please focus on reviewing the Timepoint Explorer components and tests. Use it in your browser with HTTP, Scripted and Browser (more to be added as the parent dashboards are converted to React scenes) checks in a variety of situations, including:

- Creating new checks
- Viewing existing checks
- Updating checks

Please try and use it as many scenarios as you can think of. The documentation should provide good inspiration of the different states that are available.

Unless absolutely critical this PR will not implement any additional features but feature requests and additional upgrades are being tracked in a [separate issue ticket found here](https://github.com/grafana/synthetic-monitoring-app/issues/1238).

### CODE REVIEW

File: src/components/constants.ts [hunk 1]
```diff
@@ -524,3 +524,5 @@ export const LATENCY_DESCRIPTION =
 export const STANDARD_REFRESH_INTERVAL = 1000 * 60;
 
 export const CHECK_FORM_ERROR_EVENT = `sm-check-form-error`;
+
+export const DEFAULT_FROM_TIME = '3h';
```
Comment (w1kman) [inline src/components/constants.ts]: Nit: since the file itself doesn't give any context to what "from time" actually means. A comment and or a noun in the variable name could be helpful.

File: src/scenes/components/LogsRenderer/UniqueLogLabels.utils.ts [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+
+export function uniqueLabels(log: ParsedLokiRecord<Record<string, string>, Record<string, string>>) {
+  const labels = Object.keys(log[LokiFieldNames.Labels]);
+
+  const labelsFiltered = labels
+    .filter((key) => !LOG_LABELS_COMMON.includes(key) && !LOG_LABELS_SM.includes(key))
+    .filter((key) => key !== 'msg' && !key.includes(`_extracted`) && !key.includes(`label_`))
+    .filter((key) => key !== 'time'); // it is redundant as it corresponds to the log timestamp
+
+  return labelsFiltered;
```
Comment (w1kman) [inline src/scenes/components/LogsRenderer/UniqueLogLabels.utils.ts]: - `labelsFiltered` is redundant
- Filter once

File: src/scenes/components/TimepointExplorer/CheckResultMissing.tsx [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+import { ResultUnknown } from 'scenes/components/TimepointExplorer/ResultUnknown';
+
+import { grotPropsMagnifyingGlass } from 'img';
+
+export const CheckResultMissing = () => {
+  return (
+    <ResultUnknown image={<img src={grotPropsMagnifyingGlass} alt="" />} title="Check result missing">
+      <Stack direction="column" gap={2}>
+        <Text>
+          This check did not return any results for this timepoint. Since it ran the check has been updated so we are
```
Comment (w1kman) [inline src/scenes/components/TimepointExplorer/CheckResultMissing.tsx]: I stumbled a little bit on this copy. As I'm not a native English speaker, I asked an AI, and it also thought there was some room for improvement.

I will spare you from AI-generated suggestions.

File: src/scenes/components/TimepointExplorer/ProbeResultMissing.tsx [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+        </Text>
+        .
+      </div>
+      <div>
+        We are unable to determine this probe&apos;s status when this execution was scheduled. It may be worth checking
+        if this probe reported any results for other checks that ran at{' '}
+        <TextLink href={exploreLink} external>
+          the same time
+        </TextLink>
+        .
```
Comment (w1kman) [inline src/scenes/components/TimepointExplorer/ProbeResultMissing.tsx]: AI and I feel this could be clearer and more actionable :P

**Review by w1kman — CHANGES_REQUESTED** [review]
**Code review**
Looks mint! There were a couple of texts that I felt were a little wonky.

I'm a bit concerned that we're not using `<Canvas />`, but I will get back to that (if needed) after testing the feature fully.

Nice work @ckbedwell! 💪🏻

File: src/scenes/components/TimepointExplorer/TimepointExplorerAnnotations.tsx [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+  isBeginningSection,
+  parentWidth,
+  showLabels,
+  showTooltips,
+  timepointsInRange,
+  triggerHeight,
+}: TimepointExplorerAnnotationsProps) => {
+  const { checkEvents } = useTimepointExplorerContext();
+  const styles = useStyles2(getStyles);
+  const alertsFiring = useSceneAnnotation('Alerts firing');
```
Comment (VikaCep) [inline src/scenes/components/TimepointExplorer/TimepointExplorerAnnotations.tsx]: I noticed the text is showing behind the visualization

image

File: src/scenes/components/TimepointExplorer/TimepointListErrorButton.tsx [hunk 6]
```diff
@@ -0,0 +1,38 @@
+import React from 'react';
+import { GrafanaTheme2 } from '@grafana/data';
+import { IconButton, useStyles2 } from '@grafana/ui';
+import { css } from '@emotion/css';
+
+import { useTimepointExplorerContext } from 'scenes/components/TimepointExplorer/TimepointExplorer.context';
+
+export function TimepointListErrorButton() {
```
Comment (VikaCep) [inline src/scenes/components/TimepointExplorer/TimepointListErrorButton.tsx:8]: I created a new (scripted) check and got redirected to the dashboard, where I saw the "there was an error fetching data" button as there was no data yet. After some time I did get a timepoint in the Timepoint List, but the refresh button didn't go away until I refreshed the page which was a bit confusing
image

File: src/scenes/components/TimepointExplorer/TimepointViewerActions.hooks.tsx [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+        onClick: handlePreviousTimepoint,
+      },
+      {
+        icon: 'arrow-right',
+        label: 'Next timepoint',
+        disabled: !nextTimepoint,
+        onClick: handleNextTimepoint,
+      },
+      {
+        icon: 'compass',
```
Comment (VikaCep) [inline src/scenes/components/TimepointExplorer/TimepointViewerActions.hooks.tsx]: as an alternative we could consider using the `gf-logs` icon, as the compass seems a bit more generic.

**Review by VikaCep — COMMENTED** [review]
I did a first pass today and will continue tomorrow, this looks **amazing**! 😻👏🎉

I'd appreciate some help trying to understand the following scenario from a scripted check though:

Here we see an avg duration of 388ms for `2025/08/27 20:22:00` 
image

And if we see that point in the duration by probe panel, we get a completely different result, 76.1ms
image

Should those match? Is there a bug somewhere or am I interpreting it wrong?

File: src/scenes/components/TimepointExplorer/TimepointExplorerAnnotations.tsx (hunk 5 above)
Comment (ckbedwell) [reply src/scenes/components/TimepointExplorer/TimepointExplorerAnnotations.tsx]: I couldn't decide how to handle this. If you hover over the annotation label its z-index will take it above the list entry. I figured the entry should be prioritised for view over the annotation label if you want to hover and see the result but I'm not wedded to that idea. The entry bar is larger than the label so has a larger hover target.

**Comment by ckbedwell** [comment]
> Should those match? Is there a bug somewhere or am I interpreting it wrong?

That so happens to be a bug with that panel that I am looking at fixing this cycle: https://github.com/grafana/synthetic-monitoring-app/issues/1162

With that said, I have noticed discrepancies with the logs and metrics `duartion_seconds` being reported. I am gathering enough information for the BE engineers to take a look at. Fur the purposes of the Timepoint explorer, as long as it is rendering the `duration_seconds` value in the log the component is correct.

File: src/scenes/components/TimepointExplorer/TimepointListErrorButton.tsx (hunk 6 above)
Comment (ckbedwell) [reply src/scenes/components/TimepointExplorer/TimepointListErrorButton.tsx:8]: Fixed!

File: src/scenes/components/TimepointExplorer/TimepointListEntryTooltip.tsx [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+        <span>Avg: {renderedAvgDuration}</span>
+      </div>
+    </Stack>
+  );
+};
+
+const StatusBadge = ({ status, type }: { status: TimepointStatus; type: 'uptime' | 'reachability' }) => {
+  const missingText = type === 'uptime' ? 'UNKNOWN' : 'MISSING';
+  const successText = type === 'uptime' ? 'UP' : 'SUCCESS';
+  const failureText = type === 'uptime' ? 'DOWN' : 'FAILURE';
```
Comment (VikaCep) [inline src/scenes/components/TimepointExplorer/TimepointListEntryTooltip.tsx:108]: Why not say `FAILURE` for both cases? IMO it sounds clearer as I can see `DOWN` being confused with probe infrastructure status, as suggesting the probe itself is offline
image

File: src/scenes/components/TimepointExplorer/TimepointListEntryTooltip.tsx [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+
+  switch (status) {
+    case 'missing':
+      // @ts-expect-error - it does accept gray...
+      return <Badge color="gray" text={missingText} />;
```
Comment (VikaCep) [inline src/scenes/components/TimepointExplorer/TimepointListEntryTooltip.tsx]:
```suggestion
      return <Badge color="darkgrey" text={missingText} />;
```
Looks good with `darkgrey` to remove the ts-expect-error line. Feel free to ignore.
image

File: src/scenes/components/TimepointExplorer/TimepointMinimap.tsx [hunk 10]
```diff
@@ -0,0 +1,191 @@
...
+    },
+  });
+
+  const className = isCheckCreationWithinRange
+    ? undefined
+    : isCheckCreationAfterTo
+    ? styles.beforeCreationMimic
+    : isLogsRetentionPeriodWithinTimerange
+    ? styles.outOfRetentionPeriodMimic
+    : styles.outOfRangeMimic;
```
Comment (VikaCep) [inline src/scenes/components/TimepointExplorer/TimepointMinimap.tsx]:
```suggestion
const getMinimapClassName = () => {
  if (isCheckCreationWithinRange) {
    return undefined;
  }
  if (isCheckCreationAfterTo) {
    return styles.beforeCreationMimic;
  }
  if (isLogsRetentionPeriodWithinTimerange) {
    return styles.outOfRetentionPeriodMimic;
  }
  return styles.outOfRangeMimic;
};

const className = getMinimapClassName();
```

File: src/components/PlainButton/PlainButton.tsx [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+
+    return (
+      <button {...props} className={cx(styles.button, props.className)} ref={ref}>
+        {children}
+      </button>
+    );
+  }
+);
+
+PlainButton.displayName = 'PlainButton';
```
Comment (VikaCep) [inline src/components/PlainButton/PlainButton.tsx]:
```suggestion
export const PlainButton = forwardRef<HTMLButtonElement, PlainButtonProps>(
  function PlainButton({ children, showFocusStyles = true, ...props }, ref) {
    const styles = getStyles(useTheme2(), showFocusStyles);

    return (
      <button {...props} className={cx(styles.button, props.className)} ref={ref}>
        {children}
      </button>
    );
  }
);

```

File: src/data/useLogsRetention.ts [hunk 12]
```diff
@@ -0,0 +1,22 @@
+import { config } from '@grafana/runtime';
+
+import { UnixTimestamp } from 'scenes/components/TimepointExplorer/TimepointExplorer.types';
+
+const FREE_TRIAL_RETENTION_DAYS = 14;
+const STANDARD_RETENTION_DAYS = 31;
+
+// todo: this is a temporary solution to get the logs retention period.
+// really we should get this from gcom or another endpoint
+// some clients will have custom retention periods which makes this unreliable
```
Comment (VikaCep) [inline src/data/useLogsRetention.ts:10]: So for these clients we'll show the "Out of retention" annotation incorrectly?

File: src/features/parseCheckLogs/checkLogs.types.ts [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+  service_name: string; // same as job
+  source: 'synthetic-monitoring-agent'; // might be more?
+};
+
+export type ExecutionLabelType = {
+  check_name: string; // I
+  detected_level: string; // S
+  instance: string; // I
+  job: string; // I
+};
```
Comment (VikaCep) [inline src/features/parseCheckLogs/checkLogs.types.ts]: Should those comments be cleaned up?

File: src/features/parseLokiLogs/parseLokiLogs.ts [hunk 14]
```diff
@@ -0,0 +1,54 @@
+import {
+  LokiDataFrame,
+  LokiFieldNames,
+  LokiFields,
+  ParsedLokiRecord,
+} from 'features/parseLokiLogs/parseLokiLogs.types';
+
+type FieldParser = Partial<Record<LokiFieldNames, (value: any) => any>>;
+
+// ensure to send logfmt to Loki so the line is fully parsed
```
Comment (VikaCep) [inline src/features/parseLokiLogs/parseLokiLogs.ts]: This comment confused me the first time I read it, can we improve it to something like the following? Feel free to improve it

```
// Expects structured log data from Loki. LogQL queries must include '| logfmt' 
// parser to extract individual fields before calling this function.
```

File: src/features/queryDatasources/queryMimir.ts [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+  queryType: 'range' | 'instant';
+  maxDataPoints?: number;
+}) {
+  return queryDS({
+    queries: [
+      {
+        refId,
+        expr: query,
+        range: queryType === 'range' ? true : false,
+        instant: queryType === 'instant' ? true : false,
```
Comment (VikaCep) [inline src/features/queryDatasources/queryMimir.ts]:
```suggestion
        range: queryType === 'range',
        instant: queryType === 'instant',
```

File: src/scenes/components/LogsRenderer/LogHTTPResponseTimings.tsx [hunk 16]
```diff
@@ -0,0 +1,70 @@
...
+
+  const timings = HTTP_REQUEST_ORDER.map((key) => {
+    const [start, end] = TIMING_CALCULATIONS[key];
+    const timing = getTiming(log.labels[start], log.labels[end]);
+
+    if (!Number.isNaN(timing)) {
+      return timing;
+    }
+
+    return -1;
```
Comment (VikaCep) [inline src/scenes/components/LogsRenderer/LogHTTPResponseTimings.tsx:32]: why -1?

**Comment by w1kman** [comment]
Just encountered this.

image

**Comment by w1kman** [comment]
Overview: The rightmost section has a red background, while the other sections do not. It (rightmost section) keeps being red while "scrubbing" the overview. So it seems to have more to do with the position than the actual segment.

image

**Review by w1kman — CHANGES_REQUESTED** [review]
The DOM size explodes when the Timeline Explorer enters the DOM.

During the time I've clicked around and played with it, the browser has crashed, gotten unresponsive for periods, and Grafana stopped serving data (prob unrelated). Tooltips that are lingering on the screen, is very common. 

Disclaimer: Performance bottlenecks are mainly visible to me when DevTools is open (Firefox still has performance issues with DevTools closed).
I have a 9950x, discrete GPU and 64GB of ram, what would happen if I had a company issued < random computer brand >?

**From a UI pov**, the transparent bars are straining for the eyes (my eyes at least), and they don't really align with other Grafana vizs.

I think it works and that it solves a problem. Not sure I think it's ready to be released to production.

If it went out to prod, I think that there should be a way for users to disable the Timeline Explorer without losing the possibility to view logs.

Before I can approve, I'd like to see a discussion about the potential performance impact this feature may have on our users.
I'd also like to see a background color on the bars (or a decision that transparent is the way).

Ideally, the overview component would be rendered onto a canvas, not directly to the DOM.

File: src/data/useLogsRetention.ts (hunk 12 above)
Comment (ckbedwell) [reply src/data/useLogsRetention.ts:10]: Yes... honestly I'm okay with it 😅 When someone complains we'll look at fixing it properly.

File: src/scenes/components/LogsRenderer/LogHTTPResponseTimings.tsx (hunk 16 above)
Comment (ckbedwell) [reply src/scenes/components/LogsRenderer/LogHTTPResponseTimings.tsx:32]: I suppose it could be 0?

I like -1 because it denotes type-safe undefined.

File: src/scenes/components/TimepointExplorer/TimepointListEntryTooltip.tsx (hunk 8 above)
Comment (ckbedwell) [reply src/scenes/components/TimepointExplorer/TimepointListEntryTooltip.tsx:108]: > Why not say FAILURE for both cases? IMO it sounds clearer as I can see DOWN being confused with probe infrastructure status, as suggesting the probe itself is offline

It's actually that exact argument why I differentiated them 😅 If you think in the context of your check -- you are checking your _own_ infrastructure for if it is up or down. Those UP and DOWN values directly correlate with **up**time so this felt clearer to me.

File: src/scenes/components/TimepointExplorer/TimepointListEntryTooltip.tsx (hunk 9 above)
Comment (ckbedwell) [reply src/scenes/components/TimepointExplorer/TimepointListEntryTooltip.tsx]: Never noticed `darkgrey` in the options 😅  Good spot!

**Review by VikaCep — CHANGES_REQUESTED** [review]
I really like the new changes with the filled bars and I can feel an improvement in performance 👏👏


It looks like the toggles to turn pending and firing annotations on/off is missing for some dashboards like HTTP:
image

And I still see the stuck tooltips, but as Thomas mentioned only when devtools is open, so might not be a blocker.

https://github.com/user-attachments/assets/60365861-7e43-43ec-af3d-20d672a636c7

**Review by VikaCep — CHANGES_REQUESTED** [review]
After I update a check, I see a gap in the minimap but in the timelist I have data at that time range: 

https://github.com/user-attachments/assets/fae1179d-80a0-4825-a805-ba6b4d37bf2d

Also, the "Check updated" annotation seems to be showing twice:
image

I also left a comment asking why we'd remove the pending toggle for per-check-alerts annotations as they should exist unless I'm missing something (https://github.com/grafana/synthetic-monitoring-app/pull/1122#issuecomment-3242611494)

**Review by VikaCep — APPROVED** [review]
LGTM! Amazing work here  🚀 🚀 🚀

**Review by w1kman — APPROVED** [review]
Lovely work! 💪🏻 
The UI can still be a little sluggish at times (chrome: only with devtools open). Felt way more snappy with Firefox now, than before.

Think it works smoother, looks better, and is "safer" to deploy to production with the added feat flags.
